Update UK coarse admin

nvkelso opened the pull request:
Downpayment on https://github.com/whosonfirst-data/whosonfirst-data/issues/1871.

This first PR for **United Kingdom** coarse admin:

- Existing `macroregion` features (4) 
    - <img width="100%" alt="image" src="https://github.com/whosonfirst-data/whosonfirst-data-admin-gb/assets/853051/d6cbab9d-7cd7-4353-9bdc-a35ee3726678">
    - Updated geometries from OS (default) and OSNI (default and rev-geo alt geom)
    - No property updates
    - Concordances: 
        - `gbr-os:census_code` for England, Wales, and Scotland.
        - `ni-os:census_code` Northern Ireland
        - Note, these seem different then the existing `gbr-ons:gss_code` we have, but these ones actually join with ONS data, so there seem to be a difference between `gbr-ons:gss_code` and `gbr-ons:census_code`?!
- New `region` features (186)
    - <img width="100%" alt="image" src="https://github.com/whosonfirst-data/whosonfirst-data-admin-gb/assets/853051/a2fe01b2-904d-49ea-8b1c-c52a83a94c1a">
    - All existing `county` features (plus some splits and 3 joins) get duplicated here, and the old `region` move to `macrocounty`
    - IDs on duplicated features change because of lifecycle rules around changing placetypes, but they get superced{ed} properties to link them to the still current `county` features.
    - These features function as the admin-1 and are reported that way to ISO. But rant: This matches ISO except for the London boroughs, which WOF will only track as `county`.
    - Updated geometries from OS and OSNI
    - New properties from
        - OS (name, area_description, area_type_description, census_code)
        - OSNI (lgdname, lgdcode)
    - Concordances: 
        - OS (census_code)
        - OSNI (lgdcode)
    - These are a mix of "administrative counties" and the "unitary authority" and/or "single-tier" unitary authorities / communities in England, Wales, and Scotland... and LGAs in Northern Ireland.
    - If we wanted to, we could bring in the other "historic" administrative counties here as mz:is_current = 0 with cessation date (this is the most WOF thing to do). Or we could replace the lieutenency areas in the `macrocounty` slot with "historic" counties?!
- New `macrocounty` - _was an unused slot_
     - <img width="100%" alt="image" src="https://github.com/whosonfirst-data/whosonfirst-data-admin-gb/assets/853051/66694f74-c4ee-4f59-ae69-a85364acae67">
     - The existing "lieutenancy areas" aka "ceremonial counties" (91) move from WOF `region` to `macrocounty` plus (8) new features for Northern Ireland
     - It turns out in Northern Ireland we had coterminous LGAs in the `region` slot, and those aren't "lieutenancy areas" there. But NI-OS does provide historic counties, which are _mostly_ "lieutenancy areas", so I modified for Belfast and Londonderry/Derry and noted that in the geom_src.
     - See below for discussion about these areas and "historical counties"... It's kinda odd to have a non-administrative unit in here that doesn't nest with the rest and is more about the monarchy than civil administration... But I'm not British so I'm probably missing something...
    - The IDs change because of lifecycle rules around changing placetypes, but they will get superced{ed} properties to link them.
    - Geometries: we _could_ import new geoms for England, Scotland, and Wales "lieutenancy areas" from OS (but I didn't because these seem low value); I did import modified version of OSNI data because it was missing before (see above)
    - Properties: There aren't new properties to bring in (except for Northern Ireland)
    - Concordances: These features don't have concordances as they relate to the sovereign not demographic administration or statistics. 
- New `county` (361) records as they are the admin-2 of UK
    - <img width="100%" alt="image" src="https://github.com/whosonfirst-data/whosonfirst-data-admin-gb/assets/853051/7ec6274d-c5a4-4bab-aa98-429ccb3cc8bf">
    - New "divisions" (197) of the now WOF `region` "administrative counties" (22) are new features, so new IDs. This will include features for the London boroughs, which will need to be marked coterminous and hierarchy_label 0.
    - The old "administrative counties" (22) get marked not current (replaced by "divisions"), superceded into new WOF `region` features.
    - Remaining 164 existing "single-tier" unitary authority WOF `county` features are coterminous features with the moved `region` features, no change to IDs. But they should also marked hierarchy_label = 0, but sharing any concordances and properties with the `region` for demographic mapping. We keep them as separate features (versus setting a placetype_alt on the `region` for `county) since `county` is a common instead of common-optional placetype.
    - All together, these changes unblock demographic mapping at this placetype
    - Updated geometries from OS and OSNI
    - New properties from
        - OS (name, area_description, area_type_description, census_code)
        - OSNI (lgdname, lgdcode)
    - Concordances: 
        - OS (census_code)
        - OSNI (lgdcode)
- Some 33 counties are also coterminous with London boroughs so they were also updated, too.
    - <img width="643" alt="image" src="https://github.com/whosonfirst-data/whosonfirst-data-admin-gb/assets/853051/b428a16a-f17d-4990-b16c-dcedd3c1f5b4">
    - Geometries from OS
    - Concordances from OS
    - Properties from OS

nvkelso committed:
move regions to macrocounty (lieutenancy areas)

nvkelso committed:
Revert "move regions to macrocounty (lieutenancy areas)"

This reverts commit d6db1f771131cdb1caaf93eb33dffb4708c10cb2.

nvkelso committed:
move regions to macrocounty (lieutenancy areas)

nvkelso committed:
remove coterminous reference to outdated regions, and copy concordances (lieutenancy areas)

nvkelso committed:
add macrocounty (lieutenancy areas) in Northern Ireland

nvkelso committed:
correct inception date for NI macrocounty

nvkelso committed:
update NI regions local govt districts (props, geom)

nvkelso committed:
create new NI county coterminous features (with region)

nvkelso committed:
update NI macroregion (default geom, alt geom, props)

nvkelso committed:
add NI macroregion reversegeo alt geom

nvkelso committed:
update ENG, WLS, SCT macroregion (default geom, alt geom, props)

nvkelso committed:
move the 22 admin counties in England to regions

nvkelso committed:
add 164 new county as districts in England

nvkelso committed:
update London borough and link to new coterminous county

nvkelso committed:
update county and dupe to new coterminous regions (England-few, Wales-all, Scotland-all)

nvkelso committed:
split 2 county into halves, and lifecycle 1 county because area increase (new IDs, England)

nvkelso committed:
ensure the 22 new admin counties in England are current WOF regions

nvkelso committed:
deprecate superceded county (Poole, England)

nvkelso committed:
ensure counties with new IDs or splits and new IDs are also coterminous regions (England)

nvkelso committed:
oops

nvkelso committed:
oops!

nvkelso commented:
The total number of new or modified current features is 683. The larger count of 1,012 in the diff is because of storing alternate geometries and sometimes modifying the coterminous props of a localadmin to exclude a deprecated feature (a few of those in Scotland). 

Localadmin are not otherwise modified in this PR, but localadmin in Scotland should probably all be deprecated as they are funky data that OS distributes for unknown reason as equivelant with civil parishes in England and Scotland but they are just duplicates of the WOF region-level council areas. The next PR will modify Scotland and Ireland locality and localadmin, and I'll address this issue then.

nvkelso committed:
[PR review] restore London the locality (and removing superceding with Greater London)

nvkelso committed:
[PR review] specify original source in src:geom_via when WOF lightly modified OS or NI-OS

nvkelso commented:
Related PR to adjust the existing NI-OS source to also include `concordance` usages: 

- https://github.com/whosonfirst/whosonfirst-sources/pull/210

stepps00 commented on data/101/750/367/101750367.geojson:
```diff
@@ -6,8 +6,9 @@
     "date:inception_upper":"0043-12-31",
     "edtf:cessation":"uuuu",
     "edtf:inception":"0043~",
+    "edtf:superseded":"2023-10-25",
```
I think the locality record for London should be kept as-is, looks like this record was picked up erroneously because of it's placetype_alt props?

nvkelso replied:
Oops, I'll fix.

stepps00 commented on data/115/885/724/5/1158857245.geojson:
```diff
@@ -6,17 +6,23 @@
+        "City of Westminster London Boro"
+    ],
+    "label:eng_x_preferred_placetype":[
+        "london borough"
```
borough?

nvkelso replied:
This is coming directly from the Ordnance Survey, so....?

stepps00 commented on data/115/885/724/5/1158857245.geojson:
```diff
@@ -6,17 +6,23 @@
-    "lbl:latitude":51.497991,
-    "lbl:longitude":-0.12875,
+    "label:eng_x_preferred_longname":[
+        "City of Westminster London Boro"
```
Looks odd.. should this be "borough"? Should the longname include all of this?

nvkelso replied:
This is coming directly from the Ordnance Survey, so....?

stepps00 commented on data/136/069/855/3/1360698553.geojson:
```diff
@@ -258,12 +259,14 @@
     "wof:repo":"whosonfirst-data-admin-gb",
-    "wof:superseded_by":[],
+    "wof:superseded_by":[
+        1880762161
```
Confirming that superseding work looks good.

stepps00 commented:
Overall, I think the work in this PR looks pretty good. Filenames, alt file references, property flags, names, and placetype swapping all look as expected. A few additional comments:

- Country geometry doesn't match macroregion boundaries (N Ireland), but should
- Some label values look truncated / incorrect (see comment)
- London locality should stay enabled. Actually, all records at the locality placetype should stay enabled through this PR.
- Are all concordances moved from record A to record B, when superseding takes place? Looks like they are, but commenting in case I've missed anything

nvkelso commented:
Thanks for the review! Action items for me:

- Country geometry doesn't match macroregion boundaries (N Ireland), but should
- London locality should stay enabled. Actually, all records at the locality placetype should stay enabled through this PR.

nvkelso commented:
> Are all concordances moved from record A to record B, when superseding takes place? Looks like they are, but commenting in case I've missed anything

Yes, the script (added [new PR](https://github.com/whosonfirst/whosonfirst-cookbook/pull/48)) moves the concordances around.

nvkelso commented:
Doing another round of QA after some tips from @stepps00, and documnting process here:

You can get a list of modified and added features in git from a main branch origin SHA and current PR head, like: 

> git diff --name-only --diff-filter=A acc6ffb9072429141a30df677c79196e3454c2d3..HEAD > ~/pr-changed-files-added.txt
> git diff --name-only --diff-filter=M acc6ffb9072429141a30df677c79196e3454c2d3..HEAD > ~/pr-changed-files-modified.txt

The additions will include alt geoms, for the sake of this QA pass I'll set those aside.

We can extract out the IDs and use the `wof-csv-to-feature-collection` script to gather to GeoJSON output:

> python wof-csv-to-feature-collection -p ~/path/to/whosonfirst-data-admin-gb/data/ -c /full/path/to/pr-changed-files-modified.csv -o /full/path/to/mods.geojson --infer-path --slim --slim-template external_editor

While QGIS can open GeoJSON files, larger geometry and nested property WOF records cause it to trip over itself. So we'll use OGR to convert to GeoPackage to speed things up.

Because WOF geoms can be large, configure OGR to make more memory available:

> export OGR_GEOJSON_MAX_OBJ_SIZE=500MB

Then:

> ogr2ogr /full/path/to/mods.gpkg /full/path/to/mods.geojson

Loading the result into QGIS we'll examine both by `placetype`, and by other properties like `mz:is_current`: 

Modified `macrocounty` records:

<img width="498" alt="image" src="https://github.com/whosonfirst-data/whosonfirst-data-admin-gb/assets/853051/b134640a-e78b-4531-b15d-9431c1ed25f0">

Modified `region` records show Northern Ireland stay current, but the records in England, Wales, and Scotland are all now not current (and superceded into new `macrocounty` records see below):

<img width="629" alt="image" src="https://github.com/whosonfirst-data/whosonfirst-data-admin-gb/assets/853051/3a186903-ab5d-491b-8a80-1803930a644c">

The areas with superceded `region` records gain _new_ records (new in pink, with those modified Northern Ireland records retained in brown):

<img width="629" alt="image" src="https://github.com/whosonfirst-data/whosonfirst-data-admin-gb/assets/853051/caa2ae5b-0d7f-4fa2-b8f6-857ff25def4d">

The old `region` records moved to new `macrocounty` records, with new macrocounty added in Northern Ireland as those were local government areas already so new lieutenancy areas needed to be created using NI-OS sources for historic counties:

<img width="568" alt="image" src="https://github.com/whosonfirst-data/whosonfirst-data-admin-gb/assets/853051/e57f3577-7f8b-4e7b-9160-1d266c82e91c">

Many of the old `county` records remain, but some – in yellow – are now no longer current (as they were superceded / moved into `region` features with new IDs). And note Scotland and Wales county records marked coterminous with their region. A handful of the counties are from splits / merges:

<img width="585" alt="image" src="https://github.com/whosonfirst-data/whosonfirst-data-admin-gb/assets/853051/644e3456-dfe9-4e1b-b767-ffa7acc6766f">

We add new `county` records to fill in the outdated areas, and to fill in missing London boroughs and create coterminous `county` features in Northern Ireland with the regions there. A handful of the counties are from splits / merges.

<img width="581" alt="image" src="https://github.com/whosonfirst-data/whosonfirst-data-admin-gb/assets/853051/2fefb0dd-3505-4b72-8250-5e97a6f3b697">

Now `localadmin` have shifted around. That happens when we remove a coterminous ID for an old `region` record for a lieutenency area that moved to a `macrocounty`. But since those `macrocounty` aren't really admin areas, I didn't want to mark them coterminous with the localadmin. (This is also what's happening for Bristol which is OK.) Plus I copied concordances over. This is further arguement this PR should also deprecate the `localadmin` level in Scotland since its not legit / already covered by coterminous `region` and `county` featues.

<img width="544" alt="image" src="https://github.com/whosonfirst-data/whosonfirst-data-admin-gb/assets/853051/b6b477d6-1b0e-4579-8b74-6be136c46274">

The London boroughs are all now coterminous with their county features:

<img width="540" alt="image" src="https://github.com/whosonfirst-data/whosonfirst-data-admin-gb/assets/853051/26b56594-a418-4ac7-aa3d-b4323a8baa94">

And looks like I accidentally modified 2 `locality` features in error in London and Bristol. The script had an error that didn't detect if the starting WOF ID was a `locality` feature, but the ID was included because the locality is placetype_alt of `county` or `region`. In those cases it needs to learn update the the locality (but not deprecate / supercede it) and also dupe it a new coterminous feature at the other core / common placetype level, which is our 2023 best practice in WOF: 

<img width="539" alt="image" src="https://github.com/whosonfirst-data/whosonfirst-data-admin-gb/assets/853051/8b079cca-1140-4b01-b46d-ffc7390ea7b5">

nvkelso committed:
[pr comments] remove unnecesary uuuu superseded date; remove placetype_alt of county

nvkelso committed:
[pr comments] London complicated hierarchy with counties (and updated other placetypes), and parenting

nvkelso commented:
Bristol `locality` was fine, just removing the ceremonial county (like with the admin), the feature is still current.

nvkelso committed:
[pr comments] deprecate Scotland localadmin (account for coterminous)

nvkelso committed:
fix NI bad geom (self intersections)

nvkelso commented:
Last TODO: 

- Country geometry doesn't match macroregion boundaries (N Ireland), but should

Now that the NI macrocounty geom is fixed, use it to dissolve with other macroregions for new default shape... and do same for the revgeo. Trying to do this in QGIS crashes the app reliably, so try Mapshaper instead.

nvkelso committed:
[pr comments] update GB country default geom (for NI)

nvkelso committed:
[pr comments] update GB country default and revgeo polygons

nvkelso commented:
New GB `country` revgeo:

<img width="578" alt="image" src="https://github.com/whosonfirst-data/whosonfirst-data-admin-gb/assets/853051/d0016e56-0da6-4f2f-95d7-b4e64b0520e8">

New GB `country` default polygon: 

<img width="598" alt="image" src="https://github.com/whosonfirst-data/whosonfirst-data-admin-gb/assets/853051/64543e44-115b-4e67-8625-276962900e2e">

nvkelso committed:
correct syntax

nvkelso commented:
Showing the `localadmin` in Scotland are now deprecated / not current: 

<img width="587" alt="image" src="https://github.com/whosonfirst-data/whosonfirst-data-admin-gb/assets/853051/cfe50be7-ac79-415e-92e9-6e37a6d45da8">

@stepps00 that's the last of the changes, over to you for final 👀 and PIP. Thanks!

stepps00 committed:
pip branch, rd 1

stepps00 committed:
simplified geoms

stepps00 committed:
update london locality record

stepps00 commented:
@nvkelso the most recent commits add all of the PIP work, some geometry simplification for large file sizes / property updates, and updates to the hierarchy, parent id, and controlled properties for London.

All properties, geometries, and geometries should also be valid.

I'm spending a bit more time to QA the changes and PIP work (the PIP job for this had a few hiccups), I'll comment here soon if everything looks good so this PR can be merged.

nvkelso commented:
@stepps00 Thank you! I'll poke around, too...

stepps00 committed:
pip branch, rd 2

stepps00 committed:
fix heath borough hierarchy and parent id

stepps00 committed:
pip branch, rd 3

stepps00 committed:
zero buffer for top level admin

stepps00 commented:
After reviewing the branch, I realized some features in Scotland partially PIP'd - so I manually reviewed all features in Scotland to make sure parent ids and hierarchy ids look alright.

And during this review, some region/county geometries were causing issues in QGIS, even though shapely found those geometries to be valid.. so one of the most recent commits zero-buffers all of the features from country down to county, updating geometries as needed.

Lastly, the locality of Heath was updated to fix https://github.com/whosonfirst-data/whosonfirst-data/issues/2170

I'm sure there is more work needed in GB, but we can merge this change for now and pick up more updates in future PRs..

nvkelso commented:
Thank you @stepps00!
